sdks/java/core/src/main/java/org/apache/beam/sdk/io/fs/MetadataDynamicCoder.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.fs;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.BiConsumer;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.CoderException;
+import org.apache.beam.sdk.coders.StructuredCoder;
+import org.apache.beam.sdk.transforms.SerializableBiConsumer;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+
+public class MetadataDynamicCoder extends StructuredCoder<MatchResult.Metadata> {
+
+  private static final MetadataCoder V1_CODER = MetadataCoder.of();
+
+  private List<Coder<?>> coders = new ArrayList<>();
```

Comment:
Do we want to use parallel lists? I would typically create a new pojo of {coder, getter, setter} and have a list of that instead, to avoid off by one errors if we ever modify this class, or need to add a new aspect to the dynamic coder that would warrant a new list

Reply:
Extracted into class in https://github.com/apache/beam/pull/15699/commits/420d4aa7705ac3a5766f08ab3b4247cd6c3330b6

Regrading parallel list, do you mean stream.parallel? and decode/encode in parallel? I think it won't work since we iterate the stream sequentially.

Reply:
Ah no, parallel list for me means when we have 2 or more lists, where listA.get(2) is logically associated with listB.get(2)